### PR TITLE
Sync preview state between overlays

### PR DESCRIPTION
**Pull request by UselessToucan**
Current preview implementation behaves weird in some scenarios.

Scenario 1:
1. Open direct overlay.
2. Start preview playback.
3. Open beatmapset overlay.
The preview button is stopped however I am hearing the beatmap track. This feels weird to me.

Scenatio 2:
1. Open direct overlay.
2. Open beatmapset overlay.
3. Start preview playback.
4. Close beatmapset overlay.
The beatmap track is stopped however I can still see it on my screen(as a direct overlay list item). This feels weird to me.

How did I resolve these issues:
The state of the play button was extracted to the `PlayButtonState` class. The relationship between `PlayButtonState` and `PlayButton` is Subject-Observer.

Known issues:
`IPreviewTrackOwner` related stuff like stopping current track on overlay closing is broken. I suggest to drop it completely or replace it with 'stop if not present' logic(stop the preview track if there are no related PlayButtons on the screen).

Video:
https://youtu.be/8zn3NIz0x-0

### CODE REVIEW

**Comment by smoogipoo** [comment]
I don't like this amount of passing and storing of states to a higher level. It's both very messy and easy to get into situations where the GC can't run due to not properly handling events/disposal of tracks, which wouldn't be an issue if `PlayButton` was the only store for the track.

I would like `PreviewTrackManager` to return a `PreviewTrack` which uses another track as its underlying track. `PlayButton` should retrieve the track and immediately query the `IsRunning` state to determine its initial state. This should ideally happen in bdl `load()` (with the `BeatmapSet` passed in via ctor), but that might be a big change to make, see how it feels.

**Comment by peppy** [comment]
Also you're using bindables wrong. You shouldn't be binding to the `ValueChanged` of another class. Create your own bindable and bind to it, then attach the event to your own one.

**Comment by UselessToucan** [comment]
> I would like PreviewTrackManager to return a PreviewTrack which uses another track as its underlying track. PlayButton should retrieve the track and immediately query the IsRunning state to determine its initial state. This should ideally happen in bdl load() (with the BeatmapSet passed in via ctor), but that might be a big change to make, see how it feels.

As far as I understand, this means that each `PlayButton` should have its own `TrackManagerPreviewTrack`, but `TrackManagerPreviewTrack` objects should share one `Track` whenever possible. I tried to implement it, but failed(see https://github.com/UselessToucan/osu/tree/sync_preview_state_between_overlays_v2).

File: osu.Game/Audio/PreviewTrackManager.cs [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
 using osu.Game.Beatmaps;
+using osu.Game.Overlays.Direct;
 
 namespace osu.Game.Audio
 {
     /// <summary>
     /// A central store for the retrieval of <see cref="PreviewTrack"/>s.
     /// </summary>
-    public class PreviewTrackManager : Component
+    public class PreviewTrackManager : CompositeDrawable
```
Comment (peppy) [inline osu.Game/Audio/PreviewTrackManager.cs]: why?

File: osu.Game/Audio/PreviewTrackManager.cs [hunk 2]
```diff
@@ -34,6 +38,31 @@ private void load(AudioManager audio, FrameworkConfigManager config)
...
+            if (result == null)
+            {
+                playButtonStates.Add(result = new PlayButtonState(beatmapSet));
+
+                result.StateChanged += state =>
+                {
+                    switch (state)
+                    {
+                        case PlayButtonState.PlaybackState.Playing:
+                            AddInternal(result);
```
Comment (peppy) [inline osu.Game/Audio/PreviewTrackManager.cs]: oh my god what is this

File: osu.Game/Overlays/Direct/PlayButtonState.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            set
+            {
+                if (previewWeakReference != null)
+                    previewWeakReference.SetTarget(value);
+                else
+                    previewWeakReference = new WeakReference<PreviewTrack>(value);
+            }
+        }
+
+        public BindableBool Playing { get; }
```
Comment (peppy) [inline osu.Game/Overlays/Direct/PlayButtonState.cs]: why bools not enum?

**Comment by peppy** [comment]
This is very convoluted right now. I cannot review further. I'd suggest looking at how #4092 is structured.

**Comment by peppy** [comment]
@UselessToucan if possible could you please update on status of this? Are you trying a different approach?

**Comment by UselessToucan** [comment]
This approach is dropped. I will try to apply Flyweight on a Storage level when I will have some free time.

**Pull request closed without merging**